Two errors encountered for nixd tests:

1. SQLite race condition ? Looks like nixd regression tests still randomly fail due to this issue.

```
nixd>  0# nixd::sigHandler(int) in nixd
nixd>  1# 0x00007FFFEF005D30 in /nix/store/flf14c3ibr83jsa070j25hg5gjapydhl-glibc-2.37-8/lib/libc.so.6
nixd>  2# __pthread_kill_implementation in /nix/store/flf14c3ibr83jsa070j25hg5gjapydhl-glibc-2.37-8/lib/libc.so.6
nixd>  3# __GI_raise in /nix/store/flf14c3ibr83jsa070j25hg5gjapydhl-glibc-2.37-8/lib/libc.so.6
nixd>  4# abort in /nix/store/flf14c3ibr83jsa070j25hg5gjapydhl-glibc-2.37-8/lib/libc.so.6
nixd>  5# __gnu_cxx::__verbose_terminate_handler() [clone .cold] in /nix/store/n7pvb7gdf1g6dvj7sl92i882qjl4kyx9-gcc-12.3.0-lib/lib/libstdc++.so.6
nixd>  6# __cxxabiv1::__terminate(void (*)()) in /nix/store/n7pvb7gdf1g6dvj7sl92i882qjl4kyx9-gcc-12.3.0-lib/lib/libstdc++.so.6
nixd>  7# 0x00007FFFEF288FF5 in /nix/store/n7pvb7gdf1g6dvj7sl92i882qjl4kyx9-gcc-12.3.0-lib/lib/libstdc++.so.6
nixd>  8# 0x00007FFFEF289247 in /nix/store/n7pvb7gdf1g6dvj7sl92i882qjl4kyx9-gcc-12.3.0-lib/lib/libstdc++.so.6
nixd>  9# nix::SQLiteError::throw_(sqlite3*, nix::hintformat&&) [clone .cold] in /nix/store/rbm1pzh07byv9s9fr99c2pqmw743jn9s-nix-2.16.1/lib/libnixstore.so
nixd> 10# void nix::SQLiteError::throw_<>(sqlite3*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) in /nix/store/rbm1pzh07byv9s9fr99c2pqmw743jn9s-nix-2.16.1/lib/libnixstore.so
nixd> 11# nix::LocalStore::openDB(nix::LocalStore::State&, bool) in /nix/store/rbm1pzh07byv9s9fr99c2pqmw743jn9s-nix-2.16.1/lib/libnixstore.so
nixd> 12# nix::LocalStore::LocalStore(std::map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > > const&) in /nix/store/rbm1pzh07byv9s9fr99c2pqmw743jn9s-nix-2.16.1/lib/libnixstore.so
nixd> 13# nix::openFromNonUri(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > > const&) in /nix/store/rbm1pzh07byv9s9fr99c2pqmw743jn9s-nix-2.16.1/lib/libnixstore.so
nixd> 14# nix::openStore(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > > const&) [clone .cold] in /nix/store/rbm1pzh07byv9s9fr99c2pqmw743jn9s-nix-2.16.1/lib/libnixstore.so
nixd> 15# nix::StoreCommand::createStore() in /nix/store/rbm1pzh07byv9s9fr99c2pqmw743jn9s-nix-2.16.1/lib/libnixcmd.so
nixd> 16# nix::StoreCommand::getStore() in /nix/store/rbm1pzh07byv9s9fr99c2pqmw743jn9s-nix-2.16.1/lib/libnixcmd.so
nixd> 17# nix::EvalCommand::getEvalState() in /nix/store/rbm1pzh07byv9s9fr99c2pqmw743jn9s-nix-2.16.1/lib/libnixcmd.so
nixd> 18# nixd::Server::evalInstallable(llvm::StringRef, int) in /build/source/build/lib/nixd/libnixd-lsp.so
nixd> 19# void llvm::detail::UniqueFunctionBase<void>::CallImpl<nixd::Server::updateWorkspaceVersion(llvm::StringRef)::{lambda()#1}>(void*) in /build/source/build/lib/nixd/libnixd-lsp.so
nixd> 20# nixd::Server::forkWorker(llvm::unique_function<void ()>, std::deque<std::unique_ptr<nixd::Server::Proc, std::default_delete<nixd::Server::Proc> >, std::allocator<std::unique_ptr<nixd::Server::Proc, std::default_delete<nixd::Server::Proc> > > >&) in /build/source/build/lib/nixd/libnixd-lsp.so
nixd> 21# nixd::Server::updateWorkspaceVersion(llvm::StringRef) in /build/source/build/lib/nixd/libnixd-lsp.so
nixd> 22# nixd::Server::addDocument(llvm::StringRef, llvm::StringRef, llvm::StringRef) in /build/source/build/lib/nixd/libnixd-lsp.so
nixd> 23# nixd::Server::onDocumentDidOpen(lspserver::DidOpenTextDocumentParams const&) in /build/source/build/lib/nixd/libnixd-lsp.so
nixd> 24# void llvm::detail::UniqueFunctionBase<void, llvm::json::Value>::CallImpl<lspserver::HandlerRegistry::addNotification<lspserver::DidOpenTextDocumentParams, nixd::Server>(llvm::StringLiteral, nixd::Server*, void (nixd::Server::*)(lspserver::DidOpenTextDocumentParams const&))::{lambda(llvm::json::Value)#1}>(void*, llvm::json::Value&) in /build/source/build/lib/nixd/libnixd-lsp.so
nixd> 25# lspserver::LSPServer::onNotify(llvm::StringRef, llvm::json::Value) in /build/source/build/lib/lspserver/libnixd-lspserver.so
nixd> 26# lspserver::InboundPort::dispatch(llvm::json::Value, lspserver::MessageHandler&) in /build/source/build/lib/lspserver/libnixd-lspserver.so
nixd> 27# lspserver::InboundPort::loop(lspserver::MessageHandler&) in /build/source/build/lib/lspserver/libnixd-lspserver.so
nixd> 28# main in nixd
nixd> 29# __libc_start_call_main in /nix/store/flf14c3ibr83jsa070j25hg5gjapydhl-glibc-2.37-8/lib/libc.so.6
nixd> 30# __libc_start_main_alias_2 in /nix/store/flf14c3ibr83jsa070j25hg5gjapydhl-glibc-2.37-8/lib/libc.so.6
nixd> 31# _start in nixd
```

2. Is there a full featured libc++ library for C++ 20 support on x86_64-darwin machines?

```
 > FAILED: lib/nixd/libnixd-lsp.dylib.p/src_ServerController.cpp.o
       > clang++ -Ilib/nixd/libnixd-lsp.dylib.p -Ilib/nixd -I../lib/nixd -I../lib/nixd/include -I../lib/lspserver/include -I/nix/store/yjqjnrxs1ki96h5l91fr0hzzjy7cci2v-llvm-16.0.1-dev/include -I/nix/store/zqjd500c6sa2kdmnniwmnhz0idj3cx3g-nix-2.16.1-dev/include/nix -I/nix/store/6649s0r7qndws6jm1ibrnyj58xax4f0y-boehm-gc-8.2.2-dev/include -I/private/tmp/nix-build-nixd-1.0.0.drv-0/source/build -fcolor-diagnostics -Wall -Winvalid-pch -std=gnu++20 -O3 -include /nix/store/zqjd500c6sa2kdmnniwmnhz0idj3cx3g-nix-2.16.1-dev/include/nix/config.h -std=c++2a -std=c++2a -std=c++2a -std=c++2a -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -MD -MQ lib/nixd/libnixd-lsp.dylib.p/src_ServerController.cpp.o -MF lib/nixd/libnixd-lsp.dylib.p/src_ServerController.cpp.o.d -o lib/nixd/libnixd-lsp.dylib.p/src_ServerController.cpp.o -c ../lib/nixd/src/ServerController.cpp
       > ../lib/nixd/src/ServerController.cpp:33:10: fatal error: 'ranges' file not found
       > #include <ranges>
       >          ^~~~~~~~
       > 1 error generated.
       ```

--- a/pkgs/development/tools/language-servers/nixd/default.nix
+++ b/pkgs/development/tools/language-servers/nixd/default.nix
@@ -1,0 +1,81 @@
+{ lib
+, stdenv
+, fetchFromGitHub
+, boost182
+, gtest
+, libbacktrace
+, lit
+, llvmPackages
+, meson
+, ninja
+, nix
+, nixpkgs-fmt
+, pkg-config
+}:
+
+stdenv.mkDerivation rec {
+  pname = "nixd";
+  version = "1.0.0";
+
+  src = fetchFromGitHub {
+    owner = "nix-community";
+    repo = "nixd";
+    rev = version;
+    hash = "sha256-kTDPbsQi9gzFAFkiAPF+V3yI1WBmILEnnsqdgHMqXJA=";
+  };
+
+  mesonBuildType = "release";
+
+  nativeBuildInputs = [
+    meson
+    ninja
+    pkg-config
+  ];
+
+  nativeCheckInputs = [
+    lit
+    nixpkgs-fmt
+  ];
+
+  buildInputs = [
+    libbacktrace
+    nix
+    gtest
+    boost182
+    llvmPackages.llvm
+  ];
+
+  env.CXXFLAGS = "-include ${nix.dev}/include/nix/config.h";
+
+  doCheck = true;
+
+  checkPhase = ''
+    runHook preCheck
+    dirs=(store var var/nix var/log/nix etc home)
+
+    for dir in $dirs; do
+      mkdir -p "$TMPDIR/$dir"
+    done
+
+    export NIX_STORE_DIR=$TMPDIR/store
+    export NIX_LOCALSTATE_DIR=$TMPDIR/var
+    export NIX_STATE_DIR=$TMPDIR/var/nix
+    export NIX_LOG_DIR=$TMPDIR/var/log/nix
+    export NIX_CONF_DIR=$TMPDIR/etc
+    export HOME=$TMPDIR/home
+
+    # Disable nixd regression tests, because it uses some features provided by
+    # nix, and does not correctly work in the sandbox
+    meson test --print-errorlogs server regression/nix-ast-dump
+    runHook postCheck
+  '';
+
+  meta = {
+    description = "Nix language server";
+    homepage = "https://github.com/nix-community/nixd";
+    license = lib.licenses.lgpl3Plus;
+    maintainers = with lib.maintainers; [ inclyc ];
+    platforms = lib.platforms.unix;
+    broken = stdenv.isDarwin;
+  };
+}

--- a/pkgs/top-level/all-packages.nix
+++ b/pkgs/top-level/all-packages.nix
@@ -17735,6 +17735,11 @@ with pkgs;
 
   nil = callPackage ../development/tools/language-servers/nil { };
 
+  nixd = callPackage ../development/tools/language-servers/nixd {
+    llvmPackages = llvmPackages_16;
+    nix = nixVersions.nix_2_16;
+  };
+
   nls = callPackage ../development/tools/language-servers/nls { };
 
   pylyzer = callPackage ../development/tools/language-servers/pylyzer { };

